_posts/2021-03-14-Julia-1.6-what-has-changed-since-1.0.md

```diff
@@ -0,0 +1,737 @@
+But now I think since it can be used for a subset of a particular type of exception it could well be useful for some of [`Base`'s interfaces](https://docs.julialang.org/en/v1/manual/interfaces/) exactly as described above.
+According to [JuliaHub search](https://juliahub.com/ui/CodeSearch?q=register_error_hint&u=all&t=all), right now there is just one package using it.
+ColorTypes.jl is [using it to explain some _consensual type-piracy_](https://github.com/JuliaGraphics/ColorTypes.jl/blob/bd31741d162361ebd44ed05ae532266998d9ce9f/src/error_hints.jl), where for some functions on its types you need to load another package.
+
+This is an experimental features, so it's not covered by the SemVer guarantees of the rest of the language.
+Still it's neat, and I don't expect it to go away,
+though I also don't expect it to graduate from experimental status until a bunch of people are using it.
+Which they will probably do as people start dropping support for older Julia versions.
+
+
+### A bunch of curried functions
+Julia has have ended up with a convention of providing curried methods of functions if that would be useful as the first argument for `filter`.
+For example, `filter(isequal(2), [1,2,3,2,1])` is the same is `filter(x->isequal(x, 2), [1,2,3,2,1])`.
+In particular these are boolean comparison-like functions:
+with two arguments, where the thing being compared against is the second.
+Julia 1.0 had `isequal`, `==` and `in`.
+Since then we have added:
+`isapprox`, `<`,`<=`,`>`, `>=`, `!=`, `startswith`, `endswith`, and `contains`.
+
+Aside: `contains` is argument-flipped `occursin`.
+It was a thing in Julia 0.6, but was removed in 1.0 and now has been added back.
+We added it back primarily so we could have the curried form, and to match `startswith` and `endswith`.
+
+### A ton of other new and improved standard library functions
+
+I am not going to manage to list all of them here.
+But I will list some of my standout favorites.
+
+
+[`@time`](https://github.com/JuliaLang/julia/pull/37678) now reports how much time was spent on compilation.
+This is going to help prevent people new to the language from including compilation time in their benchmarks.
+It is still better to use [BenchmarkTools.jl's](https://github.com/JuliaCI/BenchmarkTools.jl) `@btime`, since that does multiple samples.
+But now that too can report time spent on compilation.
+It's also useful for identifying if certain functions are taking ages to compile.
+Which I guess is its theoretical main point, but I think preventing people from benchmarking wrong is going to come up way more often.
+
+The experimental [`Base.@locals`](https://github.com/JuliaLang/julia/issues/29733)
+was added, which returns a dictionary of local variables.
+That one surprised me; I though being able to access a dictionary of local variables would get in the way of the optimizer, since it would prevent it from being able to optimize variables that are used for intermediate values away entirely.
```

Comment:
I think the fact that a macro is used to access means that it doesn't slow down code that doesn't use it.  The`@locals` macro when used will add code to add the variables to a `Dict` which will prevent the optimization, but since macros get expanded before the compiler does anything, it's 0 cost when not used.

Reply:
yeah, I suspect you are right.

I find it hard to reason about  since its not a real macro.
It just returns the expression `Expr(:locals)` 
Which expands to a chain of `Expr(:isdefined, :x)` in lowering.

```suggestion
That one surprised me; I though being able to access a dictionary of local variables would get in the way of the optimizer, since it would prevent it from being able to optimize variables that are used for intermediate values away entirely.
However, the way it functions is to generate code with creates and fills dictionary of with references to variables where it is used.
The dictionary is not maintained at all times.
So this shouldn't block optimizations in functions that don't use it.
```